Improve `unused_unsafe` lint

steffahn opened the pull request:
I’m going to add some motivation and explanation below, particularly pointing the changes in behavior from this PR.

_Edit:_ Looking for existing issues, looks like this PR fixes #88260.

_Edit2:_ Now also contains code that closes #90776.

rust-highfive commented:
r? @petrochenkov

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5079347546?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/f3a599222b3b6b60b1f293a17100bb946839988d/checks/5079347546/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 2000/12589
.................................................................................................... 2100/12589
.i.................................................................................................. 2200/12589
.................................................................................................... 2300/12589
.......................................................................F...F........................ 2400/12589
.................................................................................................... 2600/12589
.................................................................................................... 2700/12589
.................................................................................................... 2800/12589
.................................................................i.................................. 2900/12589
---
failures:

---- [ui] ui/consts/const-eval/heap/alloc_intrinsic_zero_sized.rs stdout ----
normalized stderr:
warning: unnecessary `unsafe` block
   |
LL |         unsafe {
LL |         unsafe {
   |         ^^^^^^ unnecessary `unsafe` block
   = note: `#[warn(unused_unsafe)]` on by default

warning: 1 warning emitted

---
To only update this specific test, also pass `--test-args consts/const-eval/heap/alloc_intrinsic_zero_sized.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/consts/const-eval/heap/alloc_intrinsic_zero_sized.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/consts/const-eval/heap/alloc_intrinsic_zero_sized/a" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/consts/const-eval/heap/alloc_intrinsic_zero_sized/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
warning: unnecessary `unsafe` block
   |
LL |         unsafe {
LL |         unsafe {
   |         ^^^^^^ unnecessary `unsafe` block
   = note: `#[warn(unused_unsafe)]` on by default

warning: 1 warning emitted



------------------------------------------


---- [ui] ui/consts/const-eval/heap/dealloc_intrinsic_zero_sized.rs stdout ----
normalized stderr:
warning: unnecessary `unsafe` block
   |
LL |         unsafe {
LL |         unsafe {
   |         ^^^^^^ unnecessary `unsafe` block
   = note: `#[warn(unused_unsafe)]` on by default

warning: 1 warning emitted

---
To only update this specific test, also pass `--test-args consts/const-eval/heap/dealloc_intrinsic_zero_sized.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/consts/const-eval/heap/dealloc_intrinsic_zero_sized.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/consts/const-eval/heap/dealloc_intrinsic_zero_sized/a" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/consts/const-eval/heap/dealloc_intrinsic_zero_sized/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
warning: unnecessary `unsafe` block
   |
LL |         unsafe {
LL |         unsafe {
   |         ^^^^^^ unnecessary `unsafe` block
   = note: `#[warn(unused_unsafe)]` on by default

warning: 1 warning emitted

```

</details>

steffahn commented:
By the way, this is the first time I’ve touched rustc source code (prior contributions of mine were standard-library only).

Main motivation: Fixes some issues with the current behavior. This PR is more-or-less completely re-implementing the unused_unsafe lint; it’s also only done in the MIR-version of the lint, the set of tests for the `-Zthir-unsafeck` version no longer succeeds (and is thus disabled, see `lint-unused-unsafe.rs`).

On current nightly, 
```rs
unsafe fn unsf() {}

fn inner_ignored() {
    unsafe {
        #[allow(unused_unsafe)]
        unsafe {
            unsf()
        }
    }
}
```

doesn’t create any warnings. This situation is not unrealistic to come by, the inner `unsafe` block could e.g. come from a macro. Actually, this PR even includes removal of one unused `unsafe` in the standard library that was missed in a similar situation. (The inner `unsafe` coming from an external macro hides the warning, too.)

The reason behind this problem is how the check currently works:
* While generating MIR, it already skips nested unsafe blocks (i.e. unsafe nested in other unsafe) so that the inner one is always the one considered unused
* To differentiate the cases of no unsafe operations inside the `unsafe` vs. a surrounding `unsafe` block, there’s some ad-hoc magic walking up the HIR to look for surrounding used `unsafe` blocks.

There’s a lot of problems with this approach besides the one presented above. E.g. the MIR-building uses checks for `unsafe_op_in_unsafe_fn` lint to decide early whether or not `unsafe` blocks in an `unsafe fn` are redundant and ought to be removed.
```rs
#[allow(unsafe_op_in_unsafe_fn)]
unsafe fn granular_disallow_op_in_unsafe_fn() {
    unsafe {
        #[deny(unsafe_op_in_unsafe_fn)]
        {
            unsf();
        }
    }
}
```
```
error: call to unsafe function is unsafe and requires unsafe block (error E0133)
  --> src/main.rs:13:13
   |
13 |             unsf();
   |             ^^^^^^ call to unsafe function
   |
note: the lint level is defined here
  --> src/main.rs:11:16
   |
11 |         #[deny(unsafe_op_in_unsafe_fn)]
   |                ^^^^^^^^^^^^^^^^^^^^^^
   = note: consult the function's documentation for information on how to avoid undefined behavior

warning: unnecessary `unsafe` block
  --> src/main.rs:10:5
   |
9  | unsafe fn granular_disallow_op_in_unsafe_fn() {
   | --------------------------------------------- because it's nested under this `unsafe` fn
10 |     unsafe {
   |     ^^^^^^ unnecessary `unsafe` block
   |
   = note: `#[warn(unused_unsafe)]` on by default

```
Here, the intermediate `unsafe` was ignored, even though it contains a unsafe operation that is not allowed to happen in an `unsafe fn` without an additional `unsafe` block.

Also closures were problematic and the workaround/algorithms used on current nightly didn’t work properly. (I skipped trying to fully understand what it was supposed to do, because this PR uses a completely different approach.)
```rs
fn nested() {
    unsafe {
        unsafe { unsf() }
    }
}
```
```
warning: unnecessary `unsafe` block
  --> src/main.rs:10:9
   |
9  |     unsafe {
   |     ------ because it's nested under this `unsafe` block
10 |         unsafe { unsf() }
   |         ^^^^^^ unnecessary `unsafe` block
   |
   = note: `#[warn(unused_unsafe)]` on by default
```

vs

```rs
fn nested() {
    let _ = || unsafe {
        let _ = || unsafe { unsf() };
    };
}
```
```
warning: unnecessary `unsafe` block
 --> src/main.rs:9:16
  |
9 |     let _ = || unsafe {
  |                ^^^^^^ unnecessary `unsafe` block
  |
  = note: `#[warn(unused_unsafe)]` on by default

warning: unnecessary `unsafe` block
  --> src/main.rs:10:20
   |
10 |         let _ = || unsafe { unsf() };
   |                    ^^^^^^ unnecessary `unsafe` block
```

*note that this warning kind-of suggests that **both** unsafe blocks are redundant*

<hr>

I also dislike the fact that it always suggests keeping the outermost `unsafe`. E.g. for
```rs
fn granularity() {
    unsafe {
        unsafe { unsf() }
        unsafe { unsf() }
        unsafe { unsf() }
    }
}
```
I prefer if `rustc` suggests removing the more-course outer-level `unsafe` instead of the fine-grained inner `unsafe` blocks, which it currently does on nightly:
```
warning: unnecessary `unsafe` block
  --> src/main.rs:10:9
   |
9  |     unsafe {
   |     ------ because it's nested under this `unsafe` block
10 |         unsafe { unsf() }
   |         ^^^^^^ unnecessary `unsafe` block
   |
   = note: `#[warn(unused_unsafe)]` on by default

warning: unnecessary `unsafe` block
  --> src/main.rs:11:9
   |
9  |     unsafe {
   |     ------ because it's nested under this `unsafe` block
10 |         unsafe { unsf() }
11 |         unsafe { unsf() }
   |         ^^^^^^ unnecessary `unsafe` block

warning: unnecessary `unsafe` block
  --> src/main.rs:12:9
   |
9  |     unsafe {
   |     ------ because it's nested under this `unsafe` block
...
12 |         unsafe { unsf() }
   |         ^^^^^^ unnecessary `unsafe` block
```

<hr>

Needless to say, this PR addresses all these points. For context, as far as my understanding goes, the main advantage of skipping inner unsafe blocks was that a test case like
```rs
fn top_level_used() {
    unsafe {
        unsf();
        unsafe { unsf() }
        unsafe { unsf() }
        unsafe { unsf() }
    }
}
```
should generate some warning because there’s redundant nested `unsafe`, however every single `unsafe` block _does_ contain some statement that uses it. Of course this PR doesn’t aim change the warnings on this kind of code example, because the current behavior, warning on all the inner `unsafe` blocks, makes sense in this case.

As mentioned, during MIR building all the unsafe blocks *are* kept now, and usage is attributed to them. The way to still generate a warning like
```
warning: unnecessary `unsafe` block
  --> src/main.rs:11:9
   |
9  |     unsafe {
   |     ------ because it's nested under this `unsafe` block
10 |         unsf();
11 |         unsafe { unsf() }
   |         ^^^^^^ unnecessary `unsafe` block
   |
   = note: `#[warn(unused_unsafe)]` on by default

warning: unnecessary `unsafe` block
  --> src/main.rs:12:9
   |
9  |     unsafe {
   |     ------ because it's nested under this `unsafe` block
...
12 |         unsafe { unsf() }
   |         ^^^^^^ unnecessary `unsafe` block

warning: unnecessary `unsafe` block
  --> src/main.rs:13:9
   |
9  |     unsafe {
   |     ------ because it's nested under this `unsafe` block
...
13 |         unsafe { unsf() }
   |         ^^^^^^ unnecessary `unsafe` block
```

in this case is by emitting a `unused_unsafe` warning for all of the `unsafe` blocks that are _within a **used** unsafe block_. 

The previous code had a little HIR traversal already anyways to collect a set of all the unsafe blocks (in order to afterwards determine which ones are unused afterwards). This PR uses such a traversal to do additional things including logic like _always_ warn for an `unsafe` block that’s inside of another **used** unsafe block. The traversal is expanded to include nested closures in the same go, this simplifies a lot of things.

The whole logic around `unsafe_op_in_unsafe_fn` is a little complicated, there’s some test cases of corner-cases in this PR. (The implementation involves differentiating between whether a used unsafe block was used exclusively by operations where `allow(unsafe_op_in_unsafe_fn)` was active.) The main goal was to make sure that code should compile successfully if all the `unused_unsafe`-warnings are addressed _simultaneously_ (by removing the respective `unsafe` blocks) no matter how complicated the patterns of `unsafe_op_in_unsafe_fn` being disallowed and allowed throughout the function are.

<hr>

One noteworthy design decision I took here: An `unsafe` block with `allow(unused_unsafe)` **is considered used** for the purposes of linting about redundante contained unsafe blocks. So while
```rs
fn granularity() {
    unsafe { //~ ERROR: unnecessary `unsafe` block
        unsafe { unsf() }
        unsafe { unsf() }
        unsafe { unsf() }
    }
}
```
warns for the outer `unsafe` block, 
```rs
fn top_level_ignored() {
    #[allow(unused_unsafe)]
    unsafe {
        #[deny(unused_unsafe)]
        {
            unsafe { unsf() } //~ ERROR: unnecessary `unsafe` block
            unsafe { unsf() } //~ ERROR: unnecessary `unsafe` block
            unsafe { unsf() } //~ ERROR: unnecessary `unsafe` block
        }
    }
}
```
warns on the inner ones.

steffahn commented:
Alright, that test failure is interesting… looks like I have some debugging to do.

steffahn commented:
Unsurprisingly, it fails for
```rs
fn f() {
    let x: [(); unsafe { size() }] = [];
}
```
too.

Probably the underlying issues here is that the [`OnlyBodies`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/nested_filter/struct.OnlyBodies.html) filter does include such non-item-like constants.

Hmm, maybe I should go all-in and try to fix #72359? 🤔, ah well, probably better as a separate thing.

steffahn commented:
Let’s see if it passes now.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5079769090?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/c14a7718dc51df02e7ec3c22ff27a38b213d6212/checks/5079769090/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
```

</details>

petrochenkov commented:
I have approximately zero experience with MIR passes, so it's probably better to find someone else for reviewing this.
r? rust-lang/compiler

steffahn commented:
I’m replacing the fix of c14a7718dc51df02e7ec3c22ff27a38b213d6212 with a different approach. The idea in c14a7718dc51df02e7ec3c22ff27a38b213d6212 was to still keep the opt-in to visiting all nested bodies, but opt-out of handling anonymous consts. For this to correlate correctly with unsafety checking (which only recurses into closures), it relies on the fact that nested bodies *only* appear in
* item-likes (i.e. items, impl items, trait-decl items)
* closures
* non-item constants (in types or array length or inline consts) via “anon-const”

The nested-bodies traversal via `OnlyBodies` skips item-likes and manually making `visit_anon_const` a no-op left only closures.

I’m adding a new commit that uses the default `None` filter again, and instead uses an opt-in approach to handle closures nonetheless.

steffahn commented:
I’m adding a fix for #90776. Feel free to tell me if this should rather be in a separate PR.

steffahn commented:
I’ve also created a way to integrate this improved `unused_unsafe` check into the `-Zthir_unsafeck` code: dc1056adeda30d9373020741844cc0915ebef8ec

I’ll keep that for a separate PR, for if this one gets accepted.

bors commented:
:umbrella: The latest upstream changes (presumably #92007) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

steffahn commented:
(The conflict is just some `use` statements, I can fix that later as a rebase might interfere with review.)

nagisa commented on src/test/ui/unsafe/rfc-2585-unsafe_op_in_unsafe_fn.mir.stderr:
```diff
@@ -76,12 +76,10 @@ LL |     unsafe {}
 LL |     unsafe { unsafe { unsf() } }
-   |     ------   ^^^^^^ unnecessary `unsafe` block
-   |     |
-   |     because it's nested under this `unsafe` block
+   |     ^^^^^^ unnecessary `unsafe` block
```
We might want to be nudging people towards larger unsafe blocks rather than to maximally small ones.

In unsafe code maintaining safety invariants will typically involve multiple independent statements, and at least some style guides suggest including both the unsafe operation itself as well any preceding statements that contribute to it being safe (e.g. a null check) into the same unsafe block.

steffahn replied:
Well, I personally much prefer more fine-grained unsafe blocks. Otherwise it’s often quite hard to identify where _exactly_ the unsafe operations actually are and whether or not one might have overlooked one of them on review. AFAIK, there’s often no clear boundary as to which operations are and which aren’t relevant to maintaining safety invariants. And often even the behavior of “unrelated” safe operations (e.g. ones that can access private fields of a data structure you’re implementing) can be highly relevant for overall soundness.

Also the behavior as implemented makes for a straightforward solution of the problem that cases such as the one fixed in f3a599222b3b6b60b1f293a17100bb946839988d weren’t reported previously, or similarly cases such as the one I described [above](https://github.com/rust-lang/rust/pull/93678#issuecomment-1030685757):


> On current nightly,
> 
> ```rust
> unsafe fn unsf() {}
> 
> fn inner_ignored() {
>     unsafe {
>         #[allow(unused_unsafe)]
>         unsafe {
>             unsf()
>         }
>     }
> }
> ```
> 
> doesn’t create any warnings.

While the case with an inner `#[allow(unused_unsafe)]` is detectable, and actually that’s what I tried to do initially when working on this PR, for the external-macro case I’d need to find out how to detect those… maybe even non-external macros, too, because it always makes little sense to get a warning that suggests removing an `unsafe` block that’s coming from a macro. OTOH, special-casing warning-behavior on macros seems weird, I don’t know if there’s precedent (for non-external macros).

<hr>

Also, considering that while `unsf()` is unsafe Rust code, wrapping it in an `unsafe` block masks the unsafety, making `unsafe { unsf() }` a safe expression, so then it would make the most sense to me that in the expression `unsafe { unsafe { unsf() } }`, the outer `unsafe` is the unused one.

Really, this `unused_unsafe` lint combines two different kinds of lints, one for an `unsafe` that’s truly unused (as in, it only contains safe code, and `unsafe { unsf() }` is safe code), and another one for detecting `unsafe` blocks within other `unsafe` blocks, where the inner `unsafe` block is ***not*** unused but still *redundant*, because it’s appearing in a context where unsafe code would have been allowed. Of course, these two lints together can – if one is not careful – create bad overall suggestions (suggesting to remove too many `unsafe` blocks). The way to resolve this taken after this PR is to consider the warning about *actually* “unused” `unsafe` blocks to take precedent, and then the warning for redundant `unsafe` blocks within other `unsafe` blocks only triggers if the containing `unsafe` block turns out not to be unused. In my opinion it makes sense if a lint called `unused_unsafe` is mainly concerned with actually unused `unsafe` blocks.

Suggesting removal of the inner `unsafe` block in the expression `unsafe { unsafe { unsf() } }` instead would give precedent to the warning about the inner `unsafe` block being redundant and then, to avoid inconsistent suggestions/warnings, act as if the `unsf()` contained within that block actually used the outer `unsafe` block.

***So… well… of course, ultimately it stays question of mainly personal opinion on which kind of warning is preferred.***

Note that the case that one unused `unsafe` block contains *multiple* other/inner `unsafe` blocks is in an edit-distance kind of way closer to code with lots of small `unsafe` blocks only, compared to code with a single big `unsafe` block only. Warning only on the outer `unsafe` block creates fewer warnings and (implicitly) suggests fewer changes to the code. A piece of code that already perfectly contains an exhaustive set of smaller `unsafe` blocks (inside of one unused bigger one) seems to me as if the author intended to write lots of small `unsafe` blocks and accidentally added a large one around them, and it does not seems all that likely that the author intended to write one large `unsafe` blocks and accidentally (perfectly) managed to also enclose every single (smaller set of) unsafe operation(s) in that large `unsafe` block with its own dedicated `unsafe` block. Note that in any case where _not_ all unsafe operations have their own smaller `unsafe` block, the lint will still effectively nudge the user to using a single large `unsafe` block anyways (because there’s no good way to suggest adding extra `unsafe` blocks and also the lint (as implemented in this PR) only goes as far as to point to a containing `unsafe` block, and doesn’t also note where else that containing `unsafe` block is actually being used (although that might actually be a reasonable future addition)). The user would need to remove the outer `unsafe` block themself to find the places where it used to be used (directly).

Certainly, on that last point, the case where only a single inner `unsafe` block appears inside a single outer `unsafe` block is a bit of a special case, because then the edit-distance for removing either the outer or the inner one becomes the same; but special-casing on that seems maybe a bit inconsistent to me.

nagisa replied:
I suspect this ultimately means that there needs to be a decision of some sort made here. I'm not entirely sure if this is T-lang question or a broader community one, though. Definitely does not feel like a question for two of us to make a call on.

nagisa replied:
Some discussion on Zulip about reaching consensus here: https://zulip-archive.rust-lang.org/stream/213817-t-lang/topic/.2393678.20unused_unsafe.20changes.html.

It seems to me overall given the discussion that it is fine to merge this as proposed.

nagisa commented on src/test/ui/span/lint-unused-unsafe.rs:
```diff
@@ -1,7 +1,11 @@
+// revisions: mir
+
+// // revisions: mir thir
+// // [thir]compile-flags: -Zthir-unsafeck
```
I suspect this may have been in place to verify some sort of possibly breaking change that has occurred between 2015 and 2018 editions. Are we sure the `thir` mode can be removed at this point?

Either way, lets not have commented-out code, or if we absolutely must have any, add a comment why it is here and what is blocking it from being uncommented.

steffahn replied:
AFAIK, `-Zthir-unsafeck` is a *new* version of the unsafety check that’s in development (though at the moment mostly inactive). I do actually have managed to integrate the new version of the `unused_unsafe` check from this PR into the thir-version, too, as mentioned in https://github.com/rust-lang/rust/pull/93678#issuecomment-1031743297, which re-activates this test case. But IMO it makes more sense to focus on the stable / non-experimental mir-version. I mean if you _want_ to review both together, I might as well add that into this PR as well, even then, we should probably finish discussing the behavior first before I’ll update both implementations accordingly.

steffahn replied:
Of course, if we keep this for a separate PR, I can (and probably should) add an explanatory comment.

nagisa replied:
Yeah, I don't think its strictly necessary for us to make sure the two implementations are in line in this same PR, nor do I think the improvements to a stable implementation should be blocked by an unstable one. I would rather not review gigantic PRs either. That said, given your explanation, I feel it is important the other implementation receives coverage still.

Maybe the most straightforward way to do so would be to split the test into two files for the time being?

steffahn replied:
Right, splitting the test for now might be a reasonable option, in order to better document that the `thir` check is unchanged so far (and in order to not make test coverage worse than before). I think I can do that tomorrow, I'll add a comment noting that it's the goal to "reunify" the tests later.

steffahn replied:
I’ve split them now

nagisa reviewed (commented):
Haven't reviewed the pass itself yet, but please do rebase. This needs a perf run.

steffahn commented:
> Haven't reviewed the pass itself yet, but please do rebase. This needs a perf run.

Actually, looks like the PR that had the conflict was reverted, and there is no merge conflict anymore.

nagisa commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 3cecf813adf356525eccc04aa6a63c7cf43502cf with merge cba9c92e82f65b5d96c1039773f6812ec659a4f7...
<!-- homu: {"type":"TryBuildStarted","head_sha":"3cecf813adf356525eccc04aa6a63c7cf43502cf","merge_sha":"cba9c92e82f65b5d96c1039773f6812ec659a4f7"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5193566342?check_suite_focus=true)
Build commit: cba9c92e82f65b5d96c1039773f6812ec659a4f7 (`cba9c92e82f65b5d96c1039773f6812ec659a4f7`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/5193566342?check_suite_focus=true"},"merge_sha":"cba9c92e82f65b5d96c1039773f6812ec659a4f7"} -->

rust-timer commented:
Queued cba9c92e82f65b5d96c1039773f6812ec659a4f7 with parent c5c610aad0a012a9228ecb83cc19e77111a52140, future [comparison URL](https://perf.rust-lang.org/compare.html?start=c5c610aad0a012a9228ecb83cc19e77111a52140&end=cba9c92e82f65b5d96c1039773f6812ec659a4f7).

rust-timer commented:
Finished benchmarking commit (cba9c92e82f65b5d96c1039773f6812ec659a4f7): [comparison url](https://perf.rust-lang.org/compare.html?start=c5c610aad0a012a9228ecb83cc19e77111a52140&end=cba9c92e82f65b5d96c1039773f6812ec659a4f7).

**Summary**: This benchmark run shows 9 relevant improvements 🎉 but 5 relevant regressions 😿 to instruction counts.
- Average relevant regression: 0.8%
- Average relevant improvement: -0.2%
- Largest improvement in instruction counts: -0.3% on `incr-unchanged` builds of `keccak debug`
- Largest regression in instruction counts: 1.3% on `incr-patched: println` builds of `webrender check`

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

nagisa commented on compiler/rustc_mir_transform/src/check_unsafety.rs:
```diff
@@ -251,61 +255,70 @@ impl<'tcx> UnsafetyChecker<'_, 'tcx> {
         &mut self,
-        violations: &[UnsafetyViolation],
-        unsafe_blocks: &[(hir::HirId, bool)],
+        violations: impl ExactSizeIterator<Item = &'a UnsafetyViolation>,
```
Style nit: AFAICT the properties of `ExactSizeIterator` are not being utilized anywhere in this function, and asking for an iterator here seems to impose both an inconvenient API (e.g. `iter::once` just above) and will also result in multiple monomorphizations of this method.

Am I missing the reason why this couldn't have stayed as a `&[...]`?

steffahn replied:
The `ExactSizeIterator` is used … actually … oh, good point! In 83025b76584f5ed597901df5c7cb1aa974c5da63, there still was one `.len()` call, heh!

Regarding using `&[]` instead, well… the point is the second `new_used_unsafe_blocks` argument that gets passed an iterator over a `HashMap`, so that can’t be a slice anymore. As long as that one’s monomorphized, the first argument doesn’t really matter anymore. On second thought, it might be technically possible to create a new, empty, `HashMap` to iterate over in the other call in `fn require_unsafe`, then avoiding multiple (as in: exactly 2) monomorphizations might become avoidable again.

steffahn replied:
I’ll keep the monomorphization for now. In fact, I think I’d like switching to an `IntoIterator` bound; makes me more convinced that the monomorphized `.for_each` calls can be optimized properly for `iter::once` and `iter::empty`.

nagisa reviewed (commented):
This overall makes a ton of sense to me. I left a bunch of style nits, but ideologically this seems sensible to me.

nagisa commented on compiler/rustc_mir_transform/src/check_unsafety.rs:
```diff
@@ -387,41 +400,109 @@ pub(crate) fn provide(providers: &mut Providers) {
+        _ => Context::Safe,
+    };
+
+    debug!(
```
As an actionable solution, I wonder if logging the span would replace the need to log the body itself?

steffahn replied:
FYI, this is only changed slightly from a previously existing call
```rs
debug!("check_unused_unsafe({:?}, body={:?}, used_unsafe={:?})", def_id, body, used_unsafe);
```
which included the `body`, too. As well as the function argument… well at least one of them.

steffahn commented on compiler/rustc_mir_transform/src/check_unsafety.rs:
```diff
@@ -251,61 +255,70 @@ impl<'tcx> UnsafetyChecker<'_, 'tcx> {
+        use UsedUnsafeBlockData::{AllAllowedInUnsafeFn, SomeDisallowedInUnsafeFn};
+
+        let update_entry = |this: &mut Self, hir_id, new_usage| {
+            match this.used_unsafe_blocks.entry(hir_id) {
+                hash_map::Entry::Occupied(mut entry) => {
+                    let entry = entry.get_mut();
+                    *entry = cmp::min(*entry, new_usage);
+                }
+                hash_map::Entry::Vacant(entry) => {
+                    entry.insert(new_usage);
+                }
+            };
```
Ah, indeed that’s shorter… technically more overhead though (doing the comparison with itself… I’m really missing a function like https://hackage.haskell.org/package/containers-0.6.5.1/docs/Data-Map-Lazy.html#v:insertWith here. Knowing that the code as-is is “more properly” doing what I’m actually intending to do, I’ll keep it as-is.

steffahn commented on compiler/rustc_mir_transform/src/check_unsafety.rs:
```diff
@@ -387,41 +400,109 @@ pub(crate) fn provide(providers: &mut Providers) {
+            // only pushes if the contained `match` doesn't `return` early
+            self.unused_unsafe.push((
+                block.hir_id,
+                match (self.context, used) {
```
Okay

steffahn commented on compiler/rustc_mir_transform/src/check_unsafety.rs:
```diff
@@ -387,41 +400,109 @@ pub(crate) fn provide(providers: &mut Providers) {
+        {
+            Context::UnsafeFn(hir_id)
+        }
+        _ => Context::Safe,
```
Using `fn_sig_by_hir_id` as suggested above would turn the `None` case into an ICE. (I don’t know whether there even are any cases where this can fail though, anyway.

steffahn commented on compiler/rustc_mir_transform/src/check_unsafety.rs:
```diff
@@ -387,41 +400,109 @@ pub(crate) fn provide(providers: &mut Providers) {
-    debug!("check_unused_unsafe({:?}, body={:?}, used_unsafe={:?})", def_id, body, used_unsafe);
 
-    let mut visitor = UnusedUnsafeVisitor { used_unsafe, unsafe_blocks };
+    let context = match tcx.hir().find(hir_id) {
+        Some(Node::Item(&hir::Item { kind: hir::ItemKind::Fn(ref sig, _, _), .. }))
+            if sig.header.unsafety == hir::Unsafety::Unsafe =>
```
Oh, looking into what `fn_sig_by_hir_id`’s implementation does, I do realize that different item-like things aren’t even handled at all currently. The code is adapted from the logic previously present in the `fn is_enclosed` function that used to exist. Turns out, I never tested code examples such as

```rs
pub struct S;

impl S {
    pub unsafe fn f() {
        unsafe {
            unsf();
        }
    }
}

unsafe fn unsf() {}
```

This code doesn’t work properly on `nightly` (it doesn’t point to the containing `unsafe fn`) and doesn’t work even worse with the code currently in this PR (it doesn’t even report any unused `unsafe`ty at all…).

I suppose using `fn_sig_by_hir_id` should fix that nicely, I’ll also add test cases.

steffahn replied:
I wouldn’t understand how `let-else` applies here.

steffahn replied:
Fixed now.

nagisa replied:
> I wouldn’t understand how `let-else` applies here.

I was thinking of something like:

```rust
let Some(sig) = fn_sig_by_hir_id(hir_id) else { bug!("cannot obtain signature item with body") };
let context = if sig.header.unsafety == hir::Unsafety::Unsafe { ... } else { ... };
```

It seems somewhat dangerous to default to `Context::Safe` when the function signature is not available.

steffahn replied:
Dang it, I just lost a previous version of this response due to my desktop environment crashing… let me re-write the main points as far as I remember.

The `hir_id` here can refer to any body owner, not just functions; this includes closures… well actually, closures are ruled-out because they’re handled differently in `fn unsafety_check_result`… and constants, e.g. const items, constants in types, the length in an in array expression, etc. Hence, failure to get a `sig` is not a bug. Failure to `.find()` the id in the first place is already handled by `fn_sig_by_hir_id` internally with a call to `bug!`.

<hr>

Note that an incorrect `Context::Safe` is not actually dangerous. This context is only used for the traversal for `unused_unsafe` warnings. Worst thing that could happen is that an unnecessary unsafe block (e.g. one that’s appearing inside of an `unsafe fn`) is not reported as such.

steffahn commented on compiler/rustc_middle/src/mir/query.rs:
```diff
@@ -114,13 +114,45 @@ pub struct UnsafetyViolation {
+    /// `unsafe` block contains no unsafe operations
+    // > ``unnecessary `unsafe` block``
+    Unused,
+    // `unsafe` block nested under another (used) `unsafe` block
```
right!

steffahn commented on compiler/rustc_middle/src/mir/query.rs:
```diff
@@ -114,13 +114,45 @@ pub struct UnsafetyViolation {
+pub enum UsedUnsafeBlockData {
+    SomeDisallowedInUnsafeFn,
+    // the HirId here indicates the first usage of the `unsafe` block
+    AllAllowedInUnsafeFn(hir::HirId),
```
okay

steffahn replied:
I kind-of derived that by reading this part in the original implementation
```rs
    // The unused unsafe blocks might not be in source order; sort them so that the unused unsafe
    // error messages are properly aligned and the issue-45107 and lint-unused-unsafe tests pass.
    unsafe_unused.sort_by_cached_key(|hir_id| tcx.hir().span(*hir_id));
```

I don’t know how true it is that HirId really corresponds in this way to source order. In any case note that the only usage of this `HirId` here is for providing one instance of where the `unsafe_op_in_unsafe_fn` lint level was defined for one of the unsafe operations in the block. (Typically, it’s going to be the same source for all usages anyways.) There is no strict need for it to be the first one, it’s more about being consistent and not providing more information than necessary. (In particular, changing the `unsafe_op_in_unsafe_fn` lint level for ***any*** of the operations would already make the `unused_unsafe` warning disappear).

steffahn replied:
I mean, just for sanity, it might make sense to somehow determine (i.e. ask somebody) whether or not `HirId` order ***does*** actually correspond (to some degree) to order of appearance in the source code. Or to order of appearance in some other immediate representation, e.g. of the AST, in case that makes a difference.

nagisa replied:
Well, the code above

```
unsafe_unused.sort_by_cached_key(|hir_id| tcx.hir().span(*hir_id));
```

is sorting the _spans_ of the HIR node, not the HirIds themselves, and those do have direct relationship with the appearance order in the source code.

It might be better if in the code below, where the comparison and logic for this invariant is implemented (that is `*entry = cmp::min(*entry, new_usage);`), the spans were compared as well?

Though my understanding of your explanation that it doesn't actually matter that this HirId actually refers to the first `unsafe` use in the source-order, is that correct? I think its okay too and the adjusted comment is clear enough, I guess, what the “first” here really means.

steffahn replied:
>sorting the *spans* of the HIR node, not the HirIds themselves

Oh… I don’t know how I missed that. Perhaps misreading it once, or even misremembering and then never double-checking.

Maybe the more systematic approach would then be to just take the first one appearing in MIR (i.e. just the first one we come across in the MIR traversal)? 🤔 

Indeed it doesn’t really matter, I would appreciate reproducible compilation results (in terms of the set of warnings) though, even in the rare edge cases where this logic actually matters, so some sort of systematic approach is good.

Right, I think I’ll change it to first appearance in MIR then, this is particularly straightforward as it’s just going to be keeping the first operation that’s encountered, making it the cheapest option, too.

steffahn replied:
I’ve changed it now

steffahn commented:
I think I now addressed or at least answered to all of your review comments so far.

bors commented:
:umbrella: The latest upstream changes (presumably #94174) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

steffahn commented:
Looks like one of those PRs that introduces using let-else in a lot of places introduced a (trivial to resolve) merge conflict. As before, I'll postpone rebasing until review is complete.

nagisa commented:
LGTM.

steffahn commented:
> LGTM.

So I can rebase?

nagisa commented:
Yeah.

steffahn commented:
Done rebasing, and CI is passing.

nagisa commented:
Could you also squash the commits together so that a more meaningful history is formed once the merge happens? In particular commits such as those dealing with rustfmt idiosyncrasies, addressing review comments and similar don't really have a lasting historical value. From what I can tell this PR could be 1 to 3 commits, one for lint changes and then the others for adjusting the test suite.

steffahn commented:
I thought about this… I’m having a hard time finding any reasonable split, besides integrating everything into a single commit. E.g. tests are already separated by being changes in different files. I guess I can just put everything into a single commit if that’s preferred over the status quo; what do you think?

nagisa commented:
A single commit would be good too. It would be awesome if its message is made to contain the text from your long explanation comment too. That way the motivation doesn't get lost when github eventually decides to pull the plug, either.

steffahn commented:
Alright, I’ve added the explanation. As you can see, the code is identical: https://github.com/rust-lang/rust/compare/99a1aa904aceb120220a46954c5d270290d4b471..8f8689fb31a4ca67b348198a082dcc81acbb89ba

So CI will still pass.

nagisa commented:
@bors r+

bors commented:
:pushpin: Commit 8f8689fb31a4ca67b348198a082dcc81acbb89ba has been approved by `nagisa`

<!-- @bors r=nagisa 8f8689fb31a4ca67b348198a082dcc81acbb89ba -->
<!-- homu: {"type":"Approved","sha":"8f8689fb31a4ca67b348198a082dcc81acbb89ba","approver":"nagisa"} -->

bors commented:
:hourglass: Testing commit 8f8689fb31a4ca67b348198a082dcc81acbb89ba with merge 45e2c2881d11324d610815bfff097e25c412199e...
<!-- homu: {"type":"BuildStarted","head_sha":"8f8689fb31a4ca67b348198a082dcc81acbb89ba","merge_sha":"45e2c2881d11324d610815bfff097e25c412199e"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5267499414?check_suite_focus=true)
Approved by: nagisa
Pushing 45e2c2881d11324d610815bfff097e25c412199e to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nagisa","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/5267499414?check_suite_focus=true"},"merge_sha":"45e2c2881d11324d610815bfff097e25c412199e"} -->

rust-timer commented:
Finished benchmarking commit (45e2c2881d11324d610815bfff097e25c412199e): [comparison url](https://perf.rust-lang.org/compare.html?start=523a1b1d388bfe82a5d0540b876d9428b6dccc9c&end=45e2c2881d11324d610815bfff097e25c412199e).

**Summary**: This benchmark run did not return any relevant results. 11 results were found to be statistically significant but too small to be relevant.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
